### PR TITLE
Allow :initializationOptions with all forms of contact in eglot-server-programs

### LINKED ISSUE #940: :initializationOptions key in eglot-server-programs does not work
* Server used:              rust-analyzer
* Emacs version:             GNU Emacs 28.1 (build 1, aarch64-apple-darwin21.1.0, NS appkit-2113.00 Version 12.0.1 (Build 21A559)) of 2022-04-05
* Operating system:          Mac OS 12.3.1
* Eglot version:             7f1f09347214ebb6e0395f3e27a04f87f544ffdd
* Eglot installation method:       straight
* Using Doom:               No

#### LSP transcript - M-x eglot-events-buffer (mandatory unless Emacs inoperable)
Not possible to provide - `eglot` is not starting.
    
#### Backtrace (mandatory, unless no error message seen or heard):

```lisp
Debugger entered--Lisp error: (invalid-slot-name "#<eglot-lsp-server eglot-lsp-server-45c85222>" :initializationOptions)
  signal(invalid-slot-name ("#<eglot-lsp-server eglot-lsp-server-45c85222>" :initializationOptions))
  slot-missing(#<eglot-lsp-server eglot-lsp-server-45c85222> :initializationOptions oset ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check"))))))
  #f(compiled-function (obj slots) "Set slots of OBJ with SLOTS which is a list of name/value pairs.\nCalled from the constructor routine." #<bytecode 0x1af713d5d27c8517>)(#<eglot-lsp-server eglot-lsp-server-45c85222> (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))))
  apply(#f(compiled-function (obj slots) "Set slots of OBJ with SLOTS which is a list of name/value pairs.\nCalled from the constructor routine." #<bytecode 0x1af713d5d27c8517>) #<eglot-lsp-server eglot-lsp-server-45c85222> (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))))
  shared-initialize(#<eglot-lsp-server eglot-lsp-server-45c85222> (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))))
  #f(compiled-function (this &optional args) "Construct the new object THIS based on ARGS.\nARGS is a property list where odd numbered elements are tags, and\neven numbered elements are the values to store in the tagged slot.\nIf you overload the `initialize-instance', there you will need to\ncall `shared-initialize' yourself, or you can call `call-next-method'\nto have this constructor called automatically.  If these steps are\nnot taken, then new objects of your class will not have their values\ndynamically set from ARGS." #<bytecode 0x15852bd85dde58c4>)(#<eglot-lsp-server eglot-lsp-server-45c85222> (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))))
  apply(#f(compiled-function (this &optional args) "Construct the new object THIS based on ARGS.\nARGS is a property list where odd numbered elements are tags, and\neven numbered elements are the values to store in the tagged slot.\nIf you overload the `initialize-instance', there you will need to\ncall `shared-initialize' yourself, or you can call `call-next-method'\nto have this constructor called automatically.  If these steps are\nnot taken, then new objects of your class will not have their values\ndynamically set from ARGS." #<bytecode 0x15852bd85dde58c4>) (#<eglot-lsp-server eglot-lsp-server-45c85222> (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check"))))))))
  #f(compiled-function (&rest cnm-args) #<bytecode 0x12388d69bf933f2a>)()
  #f(compiled-function (cl--cnm conn slots) #<bytecode 0xbce44c23eae7717>)(#f(compiled-function (&rest cnm-args) #<bytecode 0x12388d69bf933f2a>) #<eglot-lsp-server eglot-lsp-server-45c85222> (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))))
  apply(#f(compiled-function (cl--cnm conn slots) #<bytecode 0xbce44c23eae7717>) #f(compiled-function (&rest cnm-args) #<bytecode 0x12388d69bf933f2a>) (#<eglot-lsp-server eglot-lsp-server-45c85222> (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check"))))))))
  #f(compiled-function (&rest args) #<bytecode 0x193e76b4af881e1b>)(#<eglot-lsp-server eglot-lsp-server-45c85222> (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))))
  apply(#f(compiled-function (&rest args) #<bytecode 0x193e76b4af881e1b>) #<eglot-lsp-server eglot-lsp-server-45c85222> (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))))
  initialize-instance(#<eglot-lsp-server eglot-lsp-server-45c85222> (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))))
  #f(compiled-function (class &rest slots) "Default constructor for CLASS `eieio-default-superclass'.\nSLOTS are the initialization slots used by `initialize-instance'.\nThis static method is called when an object is constructed.\nIt allocates the vector used to represent an EIEIO object, and then\ncalls `initialize-instance' on that object." #<bytecode 0x1f7b812a4440be43>)(eglot-lsp-server :name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check"))))))
  apply(#f(compiled-function (class &rest slots) "Default constructor for CLASS `eieio-default-superclass'.\nSLOTS are the initialization slots used by `initialize-instance'.\nThis static method is called when an object is constructed.\nIt allocates the vector used to represent an EIEIO object, and then\ncalls `initialize-instance' on that object." #<bytecode 0x1f7b812a4440be43>) eglot-lsp-server (:name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))))
  make-instance(eglot-lsp-server :name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check"))))))
  apply(make-instance eglot-lsp-server :name "EGLOT (rustfmt/rustic-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1bcedbab61>) :request-dispatcher #f(compiled-function (server method params) #<bytecode -0xa3d2f1b321bab61>) :on-shutdown eglot--on-shutdown (:process #f(compiled-function () #<bytecode -0x14193282e46f9648>) :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))))
  eglot--connect(rustic-mode (vc . "~/projects/rustfmt/") eglot-lsp-server ("rust-analyzer" :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))) "rustic")
  eglot(rustic-mode (vc . "~/projects/rustfmt/") eglot-lsp-server ("rust-analyzer" :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))) "rustic" t)
  funcall-interactively(eglot rustic-mode (vc . "~/projects/rustfmt/") eglot-lsp-server ("rust-analyzer" :initializationOptions ((:rust-analyzer (:checkOnSave (:extraArgs ("--target-dir" "/tmp/rust-analyzer-check")))))) "rustic" t)
  call-interactively(eglot record nil)
  command-execute(eglot record)
  execute-extended-command(nil "eglot" "eglot")
  funcall-interactively(execute-extended-command nil "eglot" "eglot")
  call-interactively(execute-extended-command nil nil)
  command-execute(execute-extended-command)
```
   
#### Minimum Reproducible Example (mandatory)

1. I start Emacs GUI

2. I downloaded the latest `rust-analyzer` executable from here: https://github.com/rust-lang/rust-analyzer/releases/tag/2022-04-25 and placed it to the directory in `PATH`.
   
3. I can reproduce on any Rust file or repository, but let's take https://github.com/rust-lang/rustfmt. Go to file `rustfmt/src/config/mod.rs`. 
   
4. I have the following Eglot config in my Emacs config file:
```elisp
(use-package eglot
  :config
  (add-to-list 'eglot-server-programs '(rust-mode .
                                                  ("rust-analyzer"
                                                   :initializationOptions
                                                   ((:rust-analyzer
                                                     (:checkOnSave
                                                      (:extraArgs
                                                       ("--target-dir" "/tmp/rust-analyzer-check")))))))))
;; Rust
(use-package rustic
  :config
  (setq rustic-lsp-client 'eglot)
  (setq indent-tabs-mode nil) ; use spaces instead of tabs
    (add-hook 'eglot--managed-mode-hook (lambda () (flymake-mode -1))))
```

6. After running `M-x eglot` in any Rust file, I get the following error:

```
slot-missing: Invalid slot name: "#<eglot-lsp-server eglot-lsp-server-45ccf188>", :initializationOptions
```

I expect `eglot` to start normally, as the documentation states:
> * A list (PROGRAM [ARGS...] :initializationOptions OPTIONS),
  whereupon PROGRAM is called with ARGS as in the first option,
  and the LSP "initializationOptions" JSON object is
  constructed from OPTIONS.  If OPTIONS is a unary function, it
  is called with the server instance and should return a JSON
  object.

### PR DESCRIPTION
**Pull request by astoff**
Currently, passing `:initialiationOptions` to a server in `eglot-server-programs` is not supported when the connection is over TCP or the contact is a class.

This PR lifts this restriction this by making the ` :initializationOptions` feature orthogonal to the details of how to launch a server program.

(The initial submission of this PR phrased it differently, so note that the ensuing discussion doesn't address the description above directly.)

See als Emacs bug#57725.

Closes #940

### CODE REVIEW

**Comment by joaotavora** [comment]
I'd be more comfortable with a more surgical change such as: 
```diff
diff --git a/eglot.el b/eglot.el
index ff94d5ca5f..3b1b1a9536 100644
--- a/eglot.el
+++ b/eglot.el
@@ -796,6 +796,8 @@ treated as in `eglot-dbind'."
   :documentation
   "Represents a server. Wraps a process for LSP communication.")
 
+(cl-defmethod initialize-instance :before ((_server eglot-lsp-server) &optional args)
+  (cl-remf args :initializationOptions))
 
 ;;; Process management
 (defvar eglot--servers-by-project (make-hash-table :test #'equal)
```

This seems to do the right thing in my simple tests.

**Comment by astoff** [comment]
With this patch and setting
```
(setf (alist-get 'python-mode eglot-server-programs)
        '("jedi-language-server"
          "-v"
          :initializationOptions
          (:workspace (:symbols (:maxSymbols 200)))))
```
I get the following error
```
Error in post-command-hook (#[0 "\303\304\300\242\305#\210\306\301!\205\0r\301q\210
?\205\0\307\310\311 \")\207" [(#0) #<buffer x.py> eglot--managed-mode remove-hook post-command-hook nil buffer-live-p apply eglot--connect eglot--guess-contact] 4]): (wrong-type-argument stringp :initializationOptions)
```
This happens because, as you know
```
ELISP> (let ((data '(even :odd 1))) (cl-remf data :odd) data)
(even :odd 1)
```
So these horrible manipulations in the PR
```
            (when-let ((probe (cl-position :initializationOptions contact)))
                     (prog1 (nth (1+ probe) contact)
                       (setq contact (append (cl-subseq contact 0 probe)
                                             (cl-subseq contact (+ 2 probe)))))))
```
seem necessary.

**Comment by astoff** [comment]
Oops, the above reason for the error is wrong. It's rather something in `eglot--guess-contact`.

**Comment by astoff** [comment]
By the way, my PR is longer than your suggestion because it makes `:initializationOptions` work in combination with `:autoport` or any of the other forms an entry of `eglot-server-programs` can have (testing pending).

**Comment by joaotavora** [comment]
> Oops, the above reason for the error is wrong. It's rather something in eglot--guess-contact.

Yes, I see the error now, thanks for reporting it. 

See my new commit, which simplifies `eglot--guess-contact` to not worry about providing an interactive prompt when keywords are found in the contact.  It's just not worth it.  So, barring any more errors, I think we can fix this by actually removing complexity.

**Comment by astoff** [comment]
> So, barring any more errors, I think we can fix this by actually removing complexity.

I hope this means my version, since it prunes down the possible forms of CONTACT in `eglot-server-programs` from 6 to 5, while at the same time adding flexibility :-).

**Comment by joaotavora** [comment]
No, I'm sorry.  It's too much code for and complicates the structure of the `eglot-server-programs` docstring.  I'd rather something more surgical.  Actually I think your very first suggestion over at the Emacs bug tracker also worked fine (except for the `eglot--guess-contact` thing).

> prunes down the possible forms of CONTACT in eglot-server-programs from 6 to 5, 

It just moves a section down to the end. I actually think that complicates the docstring.

Also I don't think `:autoport` is a widely used feature.  And the docstring doesn't promise that `:autoport` works with `:initializationOptions`. Making it so doesn't seem very complicated (one just needs to prune out keywords in `eglot--inferior-bootstrap`)  But I don't think it's worth it anyway.

**Comment by astoff** [comment]
Well, it's true that this is not the minimal hack necessary to make `:initializationOptions` to work at all.

This change is slightly longer because it makes `:initializationOptions` orthogonal to the rest of the other forms a contact can have in `eglot-server-programs`. So instead of a table of 5 × 2 possibilities where 4 of them are forbidden, there are 5 × 2 combinations and all work. To me this is simpler, although of course the code is a couple lines longer.

Anyway, I'll rename the PR to reflect what it does, even though I don't expect it will be merged.

File: eglot.el [hunk 1]
```diff
@@ -1139,22 +1142,18 @@ This docstring appeases checkdoc, that's all."
...
+                      (let ((default-directory default-directory))
+                        (make-process
+                         :name readable-name
+                         :command (setq server-info (eglot--cmd contact))
+                         :connection-type 'pipe
+                         :coding 'utf-8-emacs-unix
+                         :noquery t
+                         :stderr (get-buffer-create
+                                  (format "*%s stderr*" readable-name))
+                         :file-handler t)))))))
```
Comment (astoff) [inline eglot.el:1156]: Note that this hunk, which is the biggest code change here, is just reverting things to their state before `:initializationOptions` was initially introduced.

**Comment by joaotavora** [comment]
> So instead of a table of 5 × 2 possibilities where 4 of them are forbidden, there are 5 × 2 combinations and all work. To me this is simpler, although of course the code is a couple lines longer.

In my opinion, it is not simpler.  I went to some lengths to make that list a flat list of 6 forms and I hope never to have to extend it again.

And it's not a questions of making the code "a couple of lines longer": it's about unmotivated changes.  If I don't witness an actual need for making something,I won't.  Here my need is to make `:initializationOptions` work _at all_: it was totally botched and now it's not (hopefully).

Eglot is about 4 years old.  Well-meaning ambitious contributors have come and gone.  I myself drift away often: I need the codebase to be as minimal and as stable as possible. See for example all the TRAMP related bugs: I didn't implement TRAMP support and I don't use it regularly.  There were some proposals for adding this to Eglot and I took some time to choose the simplest one.  Even that seems to cause intermittent problems I just don't have time to debug.

**Comment by astoff** [comment]
> See for example all the TRAMP related bugs: I didn't implement TRAMP support and I don't use it regularly.

The T in Tramp is for "transparent", so Eglot shouldn't do anything at all — except for translating some magic file names to the local name in the remote machine, of course. I completely agree with your approach there.

I guess I should close this after all, so thanks for your patience :-).

**Pull request closed without merging**